### PR TITLE
Added MLM pretraining options to API example

### PR DESCRIPTION
**Pull request by dpressel**
- This adds a masked language model (MLM) objective (AKA denoising auto-encoder) option
- This option cannot be used with character-compositional features yet, only BPE/WordPiece1D
- Masking follows BERT "recipe"

### CODE REVIEW

File: api-examples/pretrain-transformer-lm.py [hunk 1]
```diff
@@ -579,11 +589,28 @@ def train():
         for i, batch in enumerate(train_loader):
             x, y = batch
             inputs = {'x': x.to(args.device)}
-            labels = y.to(args.device).transpose(0, 1).contiguous()
+            labels = y.to(args.device)
+            if args.mlm:
+                # Replace 15% of tokens
+                masked_indices = torch.bernoulli(torch.full(labels.shape, 0.15)).byte()
```
Comment (blester125) [inline api-examples/pretrain-transformer-lm.py]: In BERT are all the batches totally packed and there fore we don't have to worry about the thing getting masked is in the padding region?

File: api-examples/pretrain-transformer-lm.py [hunk 2]
```diff
@@ -579,11 +589,28 @@ def train():
...
+                # Anything not masked is 0 so no loss
+                labels[~masked_indices] = 0
+                # Of the masked items, mask 80% of them with [MASK]
+                indices_replaced = torch.bernoulli(torch.full(labels.shape, 0.8)).byte() & masked_indices
+                inputs[indices_replaced] = mask_value
+                # Replace 10% of them with random words, rest preserved for auto-encoding
+                indices_random = torch.bernoulli(torch.full(labels.shape, 0.5)).byte() & masked_indices & ~indices_replaced
+                random_words = torch.randint(args.d_model, labels.shape, dtype=torch.long, device=args.device)
```
Comment (blester125) [inline api-examples/pretrain-transformer-lm.py]: From the help message `args.d_model` is `Model dimension (and embedding dsz)`. This seems to be the wrong thing to use for the high values. This seems like it should be the vocab size

Comment (blester125) [inline api-examples/pretrain-transformer-lm.py]: Are there tokens that we shouldn't include in the random words (Possible by setting the low value in the random? Things like `<PAD>` and `<GO>`?

File: api-examples/pretrain-transformer-lm.py (hunk 1 above)
Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: Yeah, this shouldnt be a concern

File: api-examples/pretrain-transformer-lm.py (hunk 2 above)
Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: Yes, this is wrong, thanks

Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: Possibly, this is a good point.  Also we know that this BPE-based masking is sub-optimal.  For now would just like to get an initial attempt in there.
